File: publisher_and_subscriber_py/publisher_and_subscriber_py/subscriber_with_custom_msg.py
import rclpy
from rclpy.node import Node
from custom_msg_and_srv.msg import EmployeeSalary

class SubscriberCustomMsg(Node):
    def __init__(self):
        super().__init__("subscriber_custom_msg")
        self.subscriber_ = self.create_subscription(
            EmployeeSalary,
            "employee",
            self.message_callback,
            10
        )
        self.get_logger().info("Subscriber Custom Msg started !")

    def message_callback(self, msg):
        self.get_logger().info(
            f"Received: {msg.first_name} {msg.last_name} - {msg.salary:.2f}"
        )

def main(args=None):
    rclpy.init(args=args)
    node = SubscriberCustomMsg()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: ros2_fundamentals/publisher_and_subscriber/src/subscriber_with_custom_msg.cpp
#include "rclcpp/rclcpp.hpp"
#include "custom_msg_and_srv/msg/employee_salary.hpp"

class SubscriberCustomMsg : public rclcpp::Node
{
public:
    SubscriberCustomMsg() : Node("subscriber_custom_msg")
    {
        subscriber_ = this->create_subscription<custom_msg_and_srv::msg::EmployeeSalary>(
            "employee", 10,
            std::bind(&SubscriberCustomMsg::messageCallback, this, std::placeholders::_1)
        );
        RCLCPP_INFO(this->get_logger(), "Subscriber Custom Msg started !");
    }

private:
    void messageCallback(const custom_msg_and_srv::msg::EmployeeSalary::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "Received: %s %s - %.2f",
            msg->first_name.c_str(), msg->last_name.c_str(), msg->salary);
    }

    rclcpp::Subscription<custom_msg_and_srv::msg::EmployeeSalary>::SharedPtr subscriber_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SubscriberCustomMsg>());
    rclcpp::shutdown();
    return 0;
}